design/dpop.md

```diff
@@ -0,0 +1,259 @@
+# OAuth 2.0 Demonstrating Proof-of-Possession at the Application Layer (DPoP)
+
+* **Status**: Notes
+* **JIRA**: TBD
+
+## Motivation
+
+DPoP is a mechanism to prevent illegal API calls from succeeding only with a stolen access token. This is done via augmenting the calls to OAuth 2.0 endpoints as well as resource servers with a special DPoP header that proves possession of a private key, and binding the issued tokens to that key. Please refer to the Resources section for more info on DPoP.    
+
+## Implementation Details
+
+DPoP is a cross-cutting concept that would affect multiple Keycloak components, namely core libraries, OAuth/OIDC endpoints, adapters, models and admin UI.
+
```

Comment:
It might be good if the adapters takes what roles. I infer that those are used as a resource server or its reverse proxy.

Reply:
To say in more detail, there are several keycloak adapters and some are for client applications while others are for reverse proxy which can do DPoP related tasks on behalf of Resource Server. I would like to know whether this design document considers only client applications or both.

Reply:
Thanks for clarifications. I think we'd focus on JavaScript and Java adapters first, since the primary use case for DPoP is SPAs.

Reply:
Thank you for your clarification.

Reply:
In regards to reverse proxy type adapters, I think supporting DPoP could be low priority since access tokens normally are not exposed to the user agent (though they might be, via a custom header).

Also supporting DPoP for proxies might be relevant if there is a requirement to connect the proxy adapter to a client that enforces DPoP (i.e. doesn't support Bearer and DPoP auth schemes simultaneously), but that should be a rare case.

Therefore, I think it should be safe to de-prioritize proxy adapter support.

Reply:
Thank you for your clarification.